Some interface changes (improvements?)

mloughran opened the pull request:
I've made quite a few changes in my fork of em-zeromq - various refactorings, code simplifications, and changes to the interface which imho make the library easier to use. I wrote them some time ago for my own use, so they may be slightly trigger happy, but I wanted to make you aware of the fork in case the changes are of interest. They're stood up fairly well to our use over the past few months.

Obviously tests are lacking: I'm happy to spend some time in that department if you wanted to accept the changes.

Cheers,
Martyn

mloughran committed:
Remove handler and replace with event emitter (experimental)

mloughran committed:
Remove un-necessarily require indirection

mloughran committed:
Remove passing handler to Context#socket

mloughran committed:
Let socket decide whether it's read/writable

mloughran committed:
Simplify, remove unnecessary abstractions

I'm pretty sure this notify_readable isn't required here… TODO: test

mloughran committed:
Simplify unbind (why the indirection?)

mloughran committed:
Simply the notify_readable code

mloughran committed:
ffi-rzmq already supports reading complete message

mloughran committed:
Added ZMQ::PAIR socket type

mloughran committed:
Allow newer eventmachine dependency

schmurfy commented on lib/em-zeromq/socket.rb:
```diff
@@ -124,25 +114,9 @@ def notify_readable
-          end
+
+        while (message = get_message)
+          emit(:message, *message)
```
Where and how are handled multi-part messages ?
Unless I miss something this only works correctly for one part messages

mloughran replied:
Sorry for disappearing from this discussion. Multi part does work fine - `get_message` returns an array of parts which are then splatted, so the interface for receiving is:

```
pull.on(:message) { |part1, part2|
  p [:part1, part1.copy_out_string]
}
```

Or if you don't know how many parts to expect:

```
pull.on(:message) { |*parts|
  p [:parts, parts.map(&:copy_out_string)]
}
```

schmurfy commented:
Overall it looks fine except for my comment above.
I am not entirely convinced by the event emitter but it does not bothers me that much.

There are some changes you made to the context/socket class I wanted to add (and in fact are but sleeping on a local branch xD).

schmurfy commented:
@mloughran can you answer the line comment I added about multi-part ?
Once this is solved I have no problem merging the pull request.

skull-squadron commented:
+1

mloughran commented:
Thanks Julien, I'll take a look at fixing the specs now

mloughran committed:
Change all specs to use event emitter

* Removed all test handlers from the specs
* Reduced timeouts so specs run much faster
* Clarified the router-dealer spec & included testing dealer identity when receiving in router

mloughran commented:
@schmurfy the tests have now been updated and are all passing. Let me know if there's anything else you need.

I'm guessing you'll want to release this as 0.4 since it breaks a few interfaces.

schmurfy commented:
I will try to have a look at this during the week end, now that eventmachine 1.0.0 is released we can switch to it too.

schmurfy commented:
Since other people complained about the current interface I guess we can give this new syntax a try, if you can just update the example in the README I will merge it.

mloughran committed:
Updated readme and example so that they work

mloughran committed:
Simple multi part example

mloughran committed:
Added a simple example for setsockopt

mloughran committed:
Added a simpler example ;)

mloughran commented:
@schmurfy I've updated the readme and example. Also added some other useful examples in case they're useful to anybody. What do you think about making the example in the readme simpler - like `simpler.rb`?

andrewvc commented:
Fantastic, looks good to me. @schmurfy your thoughts?

schmurfy commented:
no problem for me I just did not found the time to correctly review the patch before merging it, I have no problem with changing the example in the README either, go for it :)

mloughran committed:
Simplified example in readme

More complex examples can live in examples dir

mloughran commented:
@schmurfy have updated readme

andrewvc commented:
I'll check it out one very last time more fully today or tomorrow and I'll merge it. I'm not anticipating any changes, just want to make sure the examples run and no obvious bugs.

andrewvc commented:
FYI, 0.4.0 has been released including:
1. These changes
2. eventmachine dep bumped to 1.0.0

mloughran commented:
Thanks guys
